apm-agent-plugin-sdk/src/test/java/co/elastic/apm/agent/sdk/state/GlobalThreadLocalTest.java

```diff
@@ -20,12 +20,103 @@
 
 import org.junit.jupiter.api.Test;
 
+import javax.annotation.Nullable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
 public class GlobalThreadLocalTest {
 
-    GlobalThreadLocal<String> threadLocal = GlobalThreadLocal.get(GlobalThreadLocalTest.class, "test");
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
 
     @Test
     void setNullValueShouldNotThrow() {
-        threadLocal.set(null);
+        GlobalThreadLocal.get(GlobalThreadLocalTest.class, "setNullValueShouldNotThrow").set(null);
+    }
+
+    @Test
+    void testNonConstantDefaultValue() throws ExecutionException, InterruptedException {
+        final GlobalThreadLocal<Object> threadLocal = GlobalThreadLocal.get(
+            GlobalThreadLocalTest.class,
+            "testNonConstantDefaultValue",
+            new ObjectValueSupplier()
```

Comment:
[minor] may be replaced with lambda
```suggestion
            () -> new Object()
```